File: src/main/java/com/wesker926/leetcode/algorithms/p0397/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/11/19
// @description: dfs+缓存
package s1

var m = map[int]int{1: 0}

func integerReplacement(n int) int {
	if r, ok := m[n]; ok {
		return r
	}

	var cur int
	if n&1 == 0 {
		cur = integerReplacement(n>>1) + 1
	} else {
		cur = minInt(integerReplacement(n>>1), integerReplacement((n>>1)+1)) + 2
	}
	m[n] = cur
	return cur
}

func minInt(i, j int) int {
	if i <= j {
		return i
	}
	return j
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0397/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0397.s1;

import java.util.HashMap;
import java.util.Map;

/**
 * @author wesker.gh
 * @date 2021/11/19
 * @description dfs+缓存
 */
public class Solution {

    private static final Map<Integer, Integer> map = new HashMap<>();

    static {
        map.put(1, 0);
    }

    public int integerReplacement(int n) {
        if (map.containsKey(n)) {
            return map.get(n);
        }

        int cur = (n & 1) == 0 ? integerReplacement(n >> 1) + 1 : Math.min(integerReplacement(n >> 1), integerReplacement((n >> 1) + 1)) + 2;
        map.put(n, cur);
        return cur;
    }
}
